man/csvformat.1
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "CSVFORMAT" "1" "Aug 16, 2024" "2.2.0" "csvkit"
.SH NAME
csvformat \- csvformat Documentation
.SH DESCRIPTION
.sp
Convert a CSV file to a custom output format.:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
usage: csvformat [\-h] [\-d DELIMITER] [\-t] [\-q QUOTECHAR] [\-u {0,1,2,3}] [\-b]
                 [\-p ESCAPECHAR] [\-z FIELD_SIZE_LIMIT] [\-e ENCODING]
                 [\-L LOCALE] [\-S] [\-H] [\-K SKIP_LINES] [\-v] [\-l] [\-\-zero] [\-V]
                 [\-E] [\-D OUT_DELIMITER] [\-T] [\-A] [\-Q OUT_QUOTECHAR]
                 [\-U {0,1,2,3}] [\-B] [\-P OUT_ESCAPECHAR]
                 [\-M OUT_LINETERMINATOR]
                 [FILE]

Convert a CSV file to a custom output format.

positional arguments:
  FILE                  The CSV file to operate on. If omitted, will accept
                        input as piped data via STDIN.

optional arguments:
  \-h, \-\-help            show this help message and exit
  \-E, \-\-skip\-header     Do not output a header row.
  \-D OUT_DELIMITER, \-\-out\-delimiter OUT_DELIMITER
                        Delimiting character of the output file.
  \-T, \-\-out\-tabs        Specify that the output file is delimited with tabs.
                        Overrides \(dq\-D\(dq.
  \-A, \-\-out\-asv         Specify that the output file is delimited with the
                        ASCII unit separator and record separator. Overrides
                        \(dq\-T\(dq, \(dq\-D\(dq and \(dq\-M\(dq.
  \-Q OUT_QUOTECHAR, \-\-out\-quotechar OUT_QUOTECHAR
                        Character used to quote strings in the output file.
  \-U {0,1,2,3}, \-\-out\-quoting {0,1,2,3}
                        Quoting style used in the output file: 0 quote
                        minimal, 1 quote all, 2 quote non\-numeric, 3 quote
                        none.
  \-B, \-\-out\-no\-doublequote
                        Whether or not double quotes are doubled in the output
                        file.
  \-P OUT_ESCAPECHAR, \-\-out\-escapechar OUT_ESCAPECHAR
                        Character used to escape the delimiter in the output
                        file if \-\-quoting 3 (\(dqQuote None\(dq) is specified and to
                        escape the QUOTECHAR if \-\-out\-no\-doublequote is
                        specified.
  \-M OUT_LINETERMINATOR, \-\-out\-lineterminator OUT_LINETERMINATOR
                        Character used to terminate lines in the output file.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
See also: \fI\%Arguments common to all tools\fP\&.
.SH EXAMPLES
.sp
Convert a comma\-separated file to a pipe\-delimited file:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
csvformat \-D \(dq|\(dq examples/dummy.csv
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Convert to carriage return line\-endings:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
csvformat \-M $\(aq\er\(aq examples/dummy.csv
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Convert to a tab\-delimited file (TSV) with no doubling of double quotes:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
printf \(aqkey,value\en1,\(dqa \(dq\(dqquoted\(dq\(dq string\(dq\(aq | csvformat \-T \-Q🐍
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
To avoid escaping quote characters when using \fB\-\-quoting 3\fP, add \fB\-\-out\-quotechar \(dq\(dq\fP:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
csvformat \-u3 \-U3 \-Q🐍 examples/optional_quote_characters.csv
.ft P
.fi
.UNINDENT
.UNINDENT
.SH AUTHOR
Christopher Groskopf and contributors
.SH COPYRIGHT
2016, Christopher Groskopf and James McKinney
.\" Generated by docutils manpage writer.
.


CHANGELOG
0.9.2
-----

* Fix version of openpyxl.

0.9.1
-----

* Add Antonio Lima to AUTHORS.
* Add support for ndjson. (#329)
* Add missing docs for csvcut -C. (#227)
* Reorganize docs so TOC works better. (#339)
* Render docs locally with RTD theme.
* Fix header in "tricks" docs.
* Add install instructions to tutorial. (#331)
* Add killer examples to doc index. (#328)
* Reorganize doc index
* Fix broken csvkit module documentation. (#327)
* Fix version of openpyxl to work around encoding issue. (#391, #288)

0.9.0
-----

* Write missing sections of the tutorial. (#32)
* Remove -q arg from sql2csv (conflicts with common flag).
* Fix csvjoin in case where left dataset rows without all columns.
* Rewrote tutorial based on LESO data. (#324)
* Don't error in csvjson if lat/lon columns are null. (#326)
* Maintain field order in output of csvjson.
* Add unit test for json in2csv. (#77)
* Maintain key order when converting JSON into CSV. (#325.)
* Upgrade python-dateutil to version 2.2 (#304)
* Fix sorting of columns with null values. (#302)
* Added release documentation.
* Fill out short rows with null values. (#313)
* Fix unicode output for csvlook and csvstat. (#315)
* Add documentation for --zero. (#323)
* Fix Integrity error when inserting zero rows in database with csvsql. (#299)
* Add Michael Mior to AUTHORS. (#305)
* Add --count option to CSVStat.
* Implement csvformat.
* Fix bug causing CSVKitDictWriter to output 'utf-8' for blank fields.

0.8.0
-----

* Add pnaimoli to AUTHORS.
* Fix column specification in csvstat. (#236)
* Added "Tips and Tricks" documentation. (#297, #298)
* Add Espartaco Palma to AUTHORS.
* Remove unnecessary enumerate calls. (#292)
* Deprecated DBF support for Python 3+.
* Add support for Python 3.3 and 3.4 (#239)

0.7.3
-----

* Fix date handling with openpyxl > 2.0 (#285)
* Add Kristina Durivage to AUTHORS. (#243)
* Added Richard Low to AUTHORS.
* Support SQL queries "directly" on CSV files. (#276)
* Add Tasneem Raja to AUTHORS.
* Fix off-by-one error in open ended column ranges. (#238)
* Add Matt Pettis to AUTHORS.
* Add line numbers flag to csvlook (#244)
* Only install argparse for Python < 2.7. (#224)
* Add Diego Rabatone Oliveira to AUTHORS.
* Add Ryan Murphy to AUTHORS.
* Fix DBF dependency. (#270)

0.7.2
-----

* Fix CHANGELOG for release.

0.7.1
-----

* Fix homepage url in setup.py.

0.7.0
-----

* Fix XLSX datetime normalization bug. (#223)
* Add raistlin7447 to AUTHORS.
* Merged sql2csv utility (#259).
* Add Jeroen Janssens to AUTHORS.
* Validate csvsql DB connections before parsing CSVs. (#257)
* Clarify install process for Ubuntu. (#249)
* Clarify docs for --escapechar. (#242)
* Make ``import csvkit`` API compatible with ``import csv``.
* Update Travis CI link. (#258)
* Add Sébastien Fievet to AUTHORS.
* Use case-sensitive name for SQLAlchemy (#237)
* Add Travis Swicegood to AUTHORS.

0.6.1
-----

* Add Chris Rosenthal to AUTHORS.
* Fix multi-file input to csvsql. (#193)
* Passing --snifflimit=0 to disable dialect sniffing. (#190)
* Add aarcro to the AUTHORS file.
* Improve performance of csvgrep. (#204)
* Add Matt Dudys to AUTHORS.
* Add support for --skipinitialspace. (#201)
* Add Joakim Lundborg to AUTHORS.
* Add --no-inference option to in2csv and csvsql. (#206)
* Add Federico Scrinzi to AUTHORS file.
* Add --no-header-row to all tools. (#189)
* Fix csvstack blowing up on empty files. (#209)
* Add Chris Rosenthal to AUTHORS file.
* Add --db-schema option to csvsql. (#216)
* Add Shane StClair to AUTHORS file.
* Add --no-inference support to csvsort. (#222)

0.5.0
-----

* Implement geojson support in csvjson. (#159)
* Optimize writing of eight bit codecs. (#175)
* Created csvpy. (#44)
* Support --not-columns for excluding columns. (#137)
* Add Jan Schulz to AUTHORS file.
* Add Windows scripts. (#111, #176)
* csvjoin, csvsql and csvstack will no longer hold open all files. (#178)
* Added Noah Hoffman to AUTHORS.
* Make csvlook output compatible with emacs table markup. (#174)

0.4.4
-----

* Add Derek Wilson to AUTHORS.
* Add Kevin Schaul to AUTHORS.
* Add DBF support to in2csv. (#11, #160)
* Support --zero option for zero-based column indexing. (#144)
* Support mixing nulls and blanks in string columns.
* Add --blanks option to csvsql. (#149)
* Add multi-file (glob) support to csvsql. (#146)
* Add Gregory Temchenko to AUTHORS.
* Add --no-create option to csvsql. (#148)
* Add Anton Ian Sipos to AUTHORS.
* Fix broken pipe errors. (#150)

0.4.3
-----

* Begin CHANGELOG (a bit late, I'll admit).


csvkit/utilities/csvformat.py
#!/usr/bin/env python
import itertools
import sys

import agate

from csvkit.cli import QUOTING_CHOICES, CSVKitUtility, make_default_headers


class CSVFormat(CSVKitUtility):
    description = 'Convert a CSV file to a custom output format.'
    override_flags = ['blanks', 'date-format', 'datetime-format']

    def add_arguments(self):
        self.argparser.add_argument(
            '-E', '--skip-header', dest='skip_header', action='store_true',
            help='Do not output a header row.')
        self.argparser.add_argument(
            '-D', '--out-delimiter', dest='out_delimiter',
            help='Delimiting character of the output file.')
        self.argparser.add_argument(
            '-T', '--out-tabs', dest='out_tabs', action='store_true',
            help='Specify that the output file is delimited with tabs. Overrides "-D".')
        self.argparser.add_argument(
            '-A', '--out-asv', dest='out_asv', action='store_true',
            help='Specify that the output file is delimited with the ASCII unit separator and record separator. '
                 'Overrides "-T", "-D" and "-M".')
        self.argparser.add_argument(
            '-Q', '--out-quotechar', dest='out_quotechar',
            help='Character used to quote strings in the output file.')
        self.argparser.add_argument(
            '-U', '--out-quoting', dest='out_quoting', type=int, choices=QUOTING_CHOICES,
            help='Quoting style used in the output file: 0 quote minimal, 1 quote all, '
                 '2 quote non-numeric, 3 quote none.')
        self.argparser.add_argument(
            '-B', '--out-no-doublequote', dest='out_doublequote', action='store_false',
            help='Whether or not double quotes are doubled in the output file.')
        self.argparser.add_argument(
            '-P', '--out-escapechar', dest='out_escapechar',
            help='Character used to escape the delimiter in the output file if --quoting 3 ("Quote None") is '
                 'specified and to escape the QUOTECHAR if --out-no-doublequote is specified.')
        self.argparser.add_argument(
            '-M', '--out-lineterminator', dest='out_lineterminator',
            help='Character used to terminate lines in the output file.')

    def _extract_csv_writer_kwargs(self):
        kwargs = {}

        if self.args.line_numbers:
            kwargs['line_numbers'] = True

        if self.args.out_asv:
            kwargs['delimiter'] = '\x1f'
        elif self.args.out_tabs:
            kwargs['delimiter'] = '\t'
        elif self.args.out_delimiter:
            kwargs['delimiter'] = self.args.out_delimiter

        if self.args.out_asv:
            kwargs['lineterminator'] = '\x1e'
        elif self.args.out_lineterminator:
            kwargs['lineterminator'] = self.args.out_lineterminator

        for arg in ('quotechar', 'quoting', 'doublequote', 'escapechar'):
            value = getattr(self.args, f'out_{arg}')
            if value is not None:
                kwargs[arg] = value

        return kwargs

    def main(self):
        if self.additional_input_expected():
            sys.stderr.write('No input file or piped data provided. Waiting for standard input:\n')

        writer = agate.csv.writer(self.output_file, **self.writer_kwargs)

        if self.args.out_quoting == 2:
            table = agate.Table.from_csv(
                self.input_file,
                skip_lines=self.args.skip_lines,
                column_types=self.get_column_types(),
                **self.reader_kwargs,
            )

            # table.to_csv() has no option to omit the column names.
            if not self.args.skip_header:
                writer.writerow(table.column_names)

            writer.writerows(table.rows)
        else:
            reader = agate.csv.reader(self.skip_lines(), **self.reader_kwargs)
            if self.args.no_header_row:
                # Peek at a row to get the number of columns.
                _row = next(reader)
                headers = make_default_headers(len(_row))
                reader = itertools.chain([headers, _row], reader)

            if self.args.skip_header:
                next(reader)

            writer.writerows(reader)


def launch_new_instance():
    utility = CSVFormat()
    utility.run()


if __name__ == '__main__':
    launch_new_instance()
